src/lib/ScenarioEditor.svelte

```
@@ -10,20 +10,25 @@
   export let scenario: ScenarioState;
 
   function update(raw: string) {
+    let value: ScenarioState;
+    const namespace = getEvalNamespace();
+    const varNameList = namespace.map(([nm, _func]) => nm).join(",");
     try {
-      dispatch("edit", JSON.parse(raw));
+      const valueCreator = eval(`({${varNameList}}) => (${raw})`);
+      value = valueCreator(Object.fromEntries(namespace));
```

when moving `dispatch(..)` into the try-block again (at the end this time) you could use `const value = ...` instead of `let value;` and assigning the value later.

And I suppose with dispatching the event at the end, you can get `edit` events with an `undefined` payload while editing the scenario

---

the intention of moving the `dispatch(...)` out of the try/catch block is, that the try/catch block is only for detecting invalid user input aka "raw" variables content (as it ignores exceptions instead of handling them).

On the other hand if `dispatch()` throws an exception (which usually should not be the case), this is not a problem of user input but of a bug in preview.svelte which intentionally should not be ignored (but forwarded to the console/sentry/...)

----

how could the `edit` event send with `undefined`? in case of invalid "raw" this would throw an exception which jumps into the catch block which returns (without ever passing the dispatch() call)

---

you are right, I missed the return in the catch-block